### PR TITLE
[GLUTEN-8010][CORE] Don't generate native metrics if transformer don't generate relNode

### PR DESCRIPTION
**Pull request by zml1206**
## What changes were proposed in this pull request?

(Fixes: \#8010)

## How was this patch tested?

Local test.
Before PR:
image
After PR:
image

### CODE REVIEW

**Comment by zhztheplayer** [comment]
Thanks!

Though it feels like a workaround since there is no real filter node in Velox query plan? Do we have other better options?

**Comment by zml1206** [comment]
> Though it feels like a workaround since there is no real filter node in Velox query plan? Do we have other better options?

How about set metricsUpdater to `MetricsUpdater.None` when getRemainingCondition? 
In this way, there is no metrics information.
image

**Comment by zml1206** [comment]
> BTW I think one of the promising ways is to exclude the filter node from Spark query plan as well. Though I remember there were some issues with an relevant attempt.

We can't exclude the filter node, because it's output may differ from child's output.

File: backends-velox/src/main/scala/org/apache/gluten/metrics/MetricsUtil.scala [hunk 1]
```diff
@@ -54,9 +54,6 @@ object MetricsUtil extends Logging {
           MetricsUpdaterTree(
             smj.metricsUpdater(),
             Seq(treeifyMetricsUpdaters(smj.bufferedPlan), treeifyMetricsUpdaters(smj.streamedPlan)))
-        case t: TransformSupport if t.metricsUpdater() == MetricsUpdater.None =>
-          assert(t.children.size == 1, "MetricsUpdater.None can only be used on unary operator")
-          treeifyMetricsUpdaters(t.children.head)
```
Comment (zml1206) [inline backends-velox/src/main/scala/org/apache/gluten/metrics/MetricsUtil.scala:59]: It will cause operatorId inconsistencies.

**Comment by zml1206** [comment]
@zhztheplayer Can you help take a look again? Thanks.

File: backends-velox/src/main/scala/org/apache/gluten/metrics/MetricsUtil.scala [hunk 2]
```diff
@@ -219,6 +216,7 @@ object MetricsUtil extends Logging {
         })
 
     mutNode.updater match {
+      case MetricsUpdater.None =>
```
Comment (zhztheplayer) [inline backends-velox/src/main/scala/org/apache/gluten/metrics/MetricsUtil.scala:219]: Are we sure about this?

The code is likely something done by `MetricsUpdater.Todo`. I remember `MetricsUpdater.Todo` and `MetricsUpdater.None` don't share the same semantic. 

Did you check the UI? Are all the metrics still normal with this change?

Comment (zml1206) [reply backends-velox/src/main/scala/org/apache/gluten/metrics/MetricsUtil.scala:219]: It doesn't seem to make any difference whether to add it or not.
Theoretically, the size of None's operatorMetrics is 0, and a judgment safety point is added.
Also, I didn't find any UI anomalies.

Comment (zml1206) [reply backends-velox/src/main/scala/org/apache/gluten/metrics/MetricsUtil.scala:219]: I thought about it carefully later, and I should remove it, which can help us distinguish whether None is set incorrectly. If an UnsupportedOperationException is thrown later, it means that it should not be None

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/ExpandExecTransformer.scala [hunk 3]
```diff
@@ -112,13 +115,12 @@ case class ExpandExecTransformer(
 
   override protected def doTransform(context: SubstraitContext): TransformContext = {
     val childCtx = child.asInstanceOf[TransformSupport].transform(context)
-    val operatorId = context.nextOperatorId(this.nodeName)
-    if (projections == null || projections.isEmpty) {
+    if (metricsUpdater == MetricsUpdater.None) {
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/ExpandExecTransformer.scala:118]: Hi @zml1206

Can we create a new utility method, e.g, `isNoop` that can be called both by `doTransform` and `metricUpdater` in each operator? Thanks.

**Review by zhztheplayer — COMMENTED** [review]
Other code looks good to me.

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/BasicPhysicalOperatorTransformer.scala [hunk 4]
```diff
@@ -61,7 +61,9 @@ abstract class FilterExecTransformerBase(val cond: Expression, val input: SparkP
     case _ => false
   }
 
-  override def metricsUpdater(): MetricsUpdater = if (getRemainingCondition == null) {
+  override def isLoop: Boolean = getRemainingCondition == null
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/BasicPhysicalOperatorTransformer.scala]: s/isLoop/isNoop/

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/ExpandExecTransformer.scala [hunk 5]
```diff
@@ -48,7 +48,9 @@ case class ExpandExecTransformer(
     AttributeSet.fromAttributeSets(projections.flatten.map(_.references))
   }
 
-  override def metricsUpdater(): MetricsUpdater = if (projections == null || projections.isEmpty) {
+  override def isLoop: Boolean = projections == null || projections.isEmpty
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/ExpandExecTransformer.scala]: s/isLoop/isNoop/

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/SortExecTransformer.scala [hunk 6]
```diff
@@ -44,7 +44,9 @@ case class SortExecTransformer(
   @transient override lazy val metrics =
     BackendsApiManager.getMetricsApiInstance.genSortTransformerMetrics(sparkContext)
 
-  override def metricsUpdater(): MetricsUpdater = if (sortOrder == null || sortOrder.isEmpty) {
+  override def isLoop: Boolean = sortOrder == null || sortOrder.isEmpty
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/SortExecTransformer.scala]: s/isLoop/isNoop/

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/WindowExecTransformer.scala [hunk 7]
```diff
@@ -51,9 +51,9 @@ case class WindowExecTransformer(
   @transient override lazy val metrics =
     BackendsApiManager.getMetricsApiInstance.genWindowTransformerMetrics(sparkContext)
 
-  override def metricsUpdater(): MetricsUpdater = if (
-    windowExpression == null || windowExpression.isEmpty
-  ) {
+  override def isLoop: Boolean = windowExpression == null || windowExpression.isEmpty
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/WindowExecTransformer.scala]: s/isLoop/isNoop/

**Review by zhztheplayer — APPROVED** [review]
One nit. Thanks!

**Pull request merged**